fasthttpadaptor/adaptor.go

```diff
@@ -56,42 +56,169 @@ func NewFastHTTPHandler(h http.Handler) fasthttp.RequestHandler {
+
+		case <-w.flushedCh:
+			// Flush occurred before handler returned.
+			// Send the first 512 bytes and start streaming
+			// the rest of the first chunk and new data as it arrives.
+			ctx.SetStatusCode(w.StatusCode())
+			haveContentType := false
+			for k, vv := range w.Header() {
+				// Don't copy Content-Length header when
+				// streaming.
+				if k == fasthttp.HeaderContentLength {
+					continue
+				}
+
+				if k == fasthttp.HeaderContentType {
+					haveContentType = true
+				}
+
+				for _, v := range vv {
+					ctx.Response.Header.Add(k, v)
+				}
+			}
+
+			if !haveContentType {
+				// From net/http.ResponseWriter.Write:
+				// If the Header does not contain a Content-Type line, Write adds a Content-Type set
+				// to the result of passing the initial 512 bytes of written data to DetectContentType.
+				l := 512
+				b := w.firstChunk
+				if len(b) < 512 {
+					l = len(b)
+				}
+				ctx.Response.Header.Set(fasthttp.HeaderContentType, http.DetectContentType(b[:l]))
+			}
+
+			// Start streaming mode on return.
+			ctx.SetBodyStreamWriter(func(bw *bufio.Writer) {
+				// Stream the first chunk.
+				if len(w.firstChunk) > 0 {
+					_, _ = bw.Write(w.firstChunk)
+					_ = bw.Flush()
+					w.firstChunk = nil
+				}
+
+				// Stream the rest of the data that is read
+				// from the net/http handler in 32 KiB chunks.
+				//
+				// Note: Data must be manually copied in chunks
+				// as data comes in.
+				chunk := make([]byte, 32*1024)
+				for {
+					// Read net/http handler chunk.
+					n, err := w.r.Read(chunk)
+					if err != nil {
+						// Handler ended due to an io.EOF
+						// or an error occurred.
+						return
+					}
+
+					// Copy chunk to fasthttp response
+					if n > 0 {
+						_, _ = bw.Write(chunk[:n])
+						_ = bw.Flush()
+					}
+				}
+			})
+			close(w.streamingCh)
+
+		case <-w.hijackedCh:
+			// The net/http handler called w.Hijack().
+			// Copy data bidirectionally between the
+			// net/http and fasthttp connections.
+			var wg sync.WaitGroup
+			wg.Add(2)
+
+			// Note: It is safe to assume that net.Conn automatically
+			// flushes data while copying.
+			go func() {
+				defer wg.Done()
+				_, _ = io.Copy(ctx.Conn(), w.handlerConn)
+
+				// Close the fasthttp connection when
+				// the net/http connection closes
+				_ = ctx.Conn().Close()
+			}()
+			go func() {
+				defer wg.Done()
+				_, _ = io.Copy(w.handlerConn, ctx.Conn())
+				// Note: Only the net/http handler
+				// should close the connection.
+			}()
+			wg.Wait()
 		}
 	}
 }
 
 type netHTTPResponseWriter struct {
-	w          io.Writer
-	h          http.Header
-	ctx        *fasthttp.RequestCtx
-	statusCode int
+	handlerConn net.Conn
+	ctx         *fasthttp.RequestCtx
+	h           http.Header
+	r           *io.PipeReader
+	w           *io.PipeWriter
+	flushedCh   chan struct{}
+	streamingCh chan struct{}
+	hijackedCh  chan struct{}
+	firstChunk  []byte
```

Comment:
`firstChunk` is a bit of a weird name as it's only the first chunk in case of streaming, for the non-streaming case it's the full response. Maybe just call it something like `responseBody`.